SchemaDocument: Copy constructor

Hello @miloyip,

thank you for merging the Validator branch so fast.

I'm about to use it now and found a difference in behauvior of `Document` and `SchemaDocument` where I wonder if this is important. To me it seems that the copy constructor of `SchemaDocument` is implicitly deleted while `Document` maybe copied. Did I perceive this correctly? Alternatively I can of course use smart pointers, but just being curious, thus asking here :) So feel free to close it without much explanations please.

	#include <rapidjson/document.h>
	#include <rapidjson/schema.h>

    ::rapidjson::Document
    load_json_from_file(
            const std::string& path_to_file) {
        std::string json = readFileContent(path_to_file);
        ::rapidjson::Document document;
        document.Parse(json.c_str());
        return document;
    }

    ::rapidjson::SchemaDocument
    load_schema_from_file(const std::string& path_to_file) {
        ::rapidjson::Document const& document =
            load_json_from_file(path_to_file);
        return ::rapidjson::SchemaDocument(document);
    }
 
Stacktrace:

    SchemaDocument const& schema = load_schema_from_file(schema_file);
    Document const& document = load_json_from_file(path_to_file);
 
	json_parser.cpp: In function ‘rapidjson::SchemaDocument load_schema_from_file(const string&)’:
	json_parser.cpp:1397:52: error: use of deleted function ‘rapidjson::GenericSchemaDocument<rapidjson::GenericValue<rapidjson::UTF8<> > >::GenericSchemaDocument(const rapidjson::GenericSchemaDocument<rapidjson::GenericValue<rapidjson::UTF8<> > >&)’
		     return ::rapidjson::SchemaDocument(document);
		                                                ^
	In file included from json_validator.hpp:7:0,
		             from json_parser.cpp:42:
	/usr/include/rapidjson/schema.h:1304:7: note: ‘rapidjson::GenericSchemaDocument<rapidjson::GenericValue<rapidjson::UTF8<> > >::GenericSchemaDocument(const rapidjson::GenericSchemaDocument<rapidjson::GenericValue<rapidjson::UTF8<> > >&)’ is implicitly deleted because the default definition would be ill-formed:
	 class GenericSchemaDocument {
		   ^
	In file included from /usr/include/rapidjson/reader.h:23:0,
		             from /usr/include/rapidjson/document.h:20,
		             from json_parser.cpp:13:
	/usr/include/rapidjson/internal/stack.h:212:5: error: ‘rapidjson::internal::Stack<Allocator>::Stack(const rapidjson::internal::Stack<Allocator>&) [with Allocator = rapidjson::CrtAllocator]’ is private
		 Stack(const Stack&);
		 ^
	In file included from json_validator.hpp:7:0,
		             from json_parser.cpp:42:
	/usr/include/rapidjson/schema.h:1304:7: error: within this context
	 class GenericSchemaDocument {
		   ^
	In file included from /usr/include/rapidjson/reader.h:23:0,
		             from /usr/include/rapidjson/document.h:20,
		             from json_parser.cpp:13:
	/usr/include/rapidjson/internal/stack.h:212:5: error: ‘rapidjson::internal::Stack<Allocator>::Stack(const rapidjson::internal::Stack<Allocator>&) [with Allocator = rapidjson::CrtAllocator]’ is private
		 Stack(const Stack&);
		 ^
	In file included from json_validator.hpp:7:0,
		             from json_parser.cpp:42:
	/usr/include/rapidjson/schema.h:1304:7: error: within this context
	 class GenericSchemaDocument {
		   ^
	ninja: build stopped: subcommand failed.

--- a/include/rapidjson/schema.h
+++ b/include/rapidjson/schema.h
@@ -1322,7 +1322,7 @@ public:
         \param remoteProvider An optional remote schema document provider for resolving remote reference. Can be null.
         \param allocator An optional allocator instance for allocating memory. Can be null.
     */
-    GenericSchemaDocument(const ValueType& document, IRemoteSchemaDocumentProviderType* remoteProvider = 0, Allocator* allocator = 0) : 
+    GenericSchemaDocument(const ValueType& document, IRemoteSchemaDocumentProviderType* remoteProvider = 0, Allocator* allocator = 0) RAPIDJSON_NOEXCEPT : 
         remoteProvider_(remoteProvider),
         allocator_(allocator),
         ownAllocator_(),
@@ -1357,6 +1357,22 @@ public:
         schemaRef_.ShrinkToFit(); // Deallocate all memory for ref
     }
 
+#if RAPIDJSON_HAS_CXX11_RVALUE_REFS
+    //! Move constructor in C++11
+    GenericSchemaDocument(GenericSchemaDocument&& rhs) RAPIDJSON_NOEXCEPT :
+        remoteProvider_(rhs.remoteProvider_),
+        allocator_(rhs.allocator_),
+        ownAllocator_(rhs.ownAllocator_),
+        root_(rhs.root_),
+        schemaMap_(std::move(rhs.schemaMap_)),
+        schemaRef_(std::move(rhs.schemaRef_))
+    {
+        rhs.remoteProvider_ = 0;
+        rhs.allocator_ = 0;
+        rhs.ownAllocator_ = 0;
+    }
+#endif
+
     //! Destructor
     ~GenericSchemaDocument() {
         while (!schemaMap_.Empty())
@@ -1369,6 +1385,11 @@ public:
     const SchemaType& GetRoot() const { return *root_; }
 
 private:
+    //! Prohibit copying
+    GenericSchemaDocument(const GenericSchemaDocument&);
+    //! Prohibit assignment
+    GenericSchemaDocument& operator=(const GenericSchemaDocument&);
+
     struct SchemaRefEntry {
         SchemaRefEntry(const PointerType& s, const PointerType& t, const SchemaType** outSchema, Allocator *allocator) : source(s, allocator), target(t, allocator), schema(outSchema) {}
         PointerType source;

--- a/test/unittest/schematest.cpp
+++ b/test/unittest/schematest.cpp
@@ -1152,6 +1152,24 @@ TEST(SchemaValidatingWriter, Simple) {
     EXPECT_TRUE(validator.GetInvalidDocumentPointer() == SchemaDocument::PointerType(""));
 }
 
+#if RAPIDJSON_HAS_CXX11_RVALUE_REFS
+
+static SchemaDocument ReturnSchemaDocument() {
+    Document sd;
+    sd.Parse("{ \"type\": [\"number\", \"string\"] }");
+    SchemaDocument s(sd);
+    return s;
+}
+
+TEST(Schema, Issue552) {
+    SchemaDocument s = ReturnSchemaDocument();
+    VALIDATE(s, "42", true);
+    VALIDATE(s, "\"I'm a string\"", true);
+    VALIDATE(s, "{ \"an\": [ \"arbitrarily\", \"nested\" ], \"data\": \"structure\" }", true);
+}
+
+#endif // RAPIDJSON_HAS_CXX11_RVALUE_REFS
+
 #ifdef __clang__
 RAPIDJSON_DIAG_POP
 #endif

--- a/test/unittest/schematest.cpp
+++ b/test/unittest/schematest.cpp
@@ -1164,8 +1164,8 @@ static SchemaDocument ReturnSchemaDocument() {
 TEST(Schema, Issue552) {
     SchemaDocument s = ReturnSchemaDocument();
     VALIDATE(s, "42", true);
-    VALIDATE(s, "\"I'm a string\"", true);
-    VALIDATE(s, "{ \"an\": [ \"arbitrarily\", \"nested\" ], \"data\": \"structure\" }", true);
+    VALIDATE(s, "\"Life, the universe, and everything\"", true);
+    INVALIDATE(s, "[\"Life\", \"the universe\", \"and everything\"]", "", "type", "");
 }
 
 #endif // RAPIDJSON_HAS_CXX11_RVALUE_REFS

